lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
         return xs, ys, zs
 
 
-def _zalpha(colors, zs):
+def _zalpha(colors, zs, _dscl=None):
     """Modify the alphas of the color list according to depth."""
-    # FIXME: This only works well if the points for *zs* are well-spaced
-    #        in all three dimensions. Otherwise, at certain orientations,
-    #        the min and max zs are very close together.
-    #        Should really normalize against the viewing depth.
+    # Return empty if given empty
     if len(colors) == 0 or len(zs) == 0:
         return np.zeros((0, 4))
-    norm = Normalize(min(zs), max(zs))
-    sats = 1 - norm(zs) * 0.7
+
+    if _dscl is None:
+        # This only works well if the points for *zs* are well-spaced
+        # in all three dimensions. Otherwise, at certain orientations,
+        # the min and max zs are very close together.
+        # Should really normalize against the viewing depth.
+
+        # Normalize the z-depths to the range 0 - 1
+        norm = Normalize(min(zs), max(zs))
+
+        # Generate alpha multipliers using the normalized z-depths so that
+        # closer points
+        # are opaque and the furthest points are still visible, but transparent
+        sats = 1 - norm(zs) * 0.7
+
+    else:
+        # Improved normalization using a scale value derived from the XYZ
+        # limits of the plot
+
+        # Solid near, transparent far, solid default
+        if _dscl == 0:
+            sats = np.ones_like(zs)
+        else:
+            sats = np.clip(1 - (zs - min(zs)) / _dscl, 0.3, 1)
+        # print(_dscl)
+
+        # Solid near, transparent far, transparent default
+        # sats = np.clip((max(zs)-zs)/(_dscl+1e-8), 0.3, 1)
+
+        # Transparent near, solid far, solid default
+        # sats = np.clip(1-(max(zs)-zs)/(_dscl+1e-8), 0.3, 1)
+
+        # Transparent near, solid far, transparent default
+        # sats = np.clip((zs-min(zs))/(_dscl+1e-8), 0.3, 1)
```

Comment:
Yes, not sure what `Solid near, transparent far, solid default` means, especially with these different options.

Reply:
I don't think we've reached a conclusion on whether it's worthwhile implementing all the different modes, or just the one that approximates the previous behavior. Either way, here's a cleaned-up version of the function that improves on the comments:

```

def _zalpha(colors, zs, _data_scale=None, shading_mode=0):
    """Modify the alphas of the color list according to depth."""
    # Return empty if given empty
    if len(colors) == 0 or len(zs) == 0:
        return np.zeros((0, 4))

    if _data_scale is None:
        # This only works well if the points for *zs* are well-spaced
        # in all three dimensions. Otherwise, at certain orientations,
        # the min and max zs are very close together.
        # Should really normalize against the viewing depth.

        # Normalize the z-depths to the range 0 - 1
        norm = Normalize(min(zs), max(zs))

        # Generate alpha multipliers using the normalized z-depths so that
        # closer points are opaque and the furthest points are still visible,
        # but transparent
        sats = 1 - norm(zs) * 0.7

    else:
        # Improved normalization using a scale value derived from the XYZ
        # limits of the plot

        if _data_scale == 0:
            # Don't scale the alpha values since we have no valid
            # data scale for reference
            sats = np.ones_like(zs)

        else:
            if shading_mode == 0:
                # This is the mode that most closely matches the behavior
                # of when _data_scale=None, and legacy code

                # Shallower points have an increasingly solid appearance
                # Deeper points have an increasingly transparent appearance
                # Points with uniform depth have a solid appearance
                sats = np.clip(1 - (zs - min(zs)) / _data_scale, 0.3, 1)

            elif shading_mode == 1:
                # Shallower points have an increasingly solid appearance
                # Deeper points have an increasingly transparent appearance
                # Points with uniform depth have a transparent appearance
                sats = np.clip((max(zs) - zs) / _data_scale, 0.3, 1)

            elif shading_mode == 2:
                # Shallower points have an increasingly transparent appearance
                # Deeper points have an increasingly solid appearance
                # Points with uniform depth have a solid appearance
                sats = np.clip(1 - (max(zs) - zs) / _data_scale, 0.3, 1)

            elif shading_mode == 3:
                # Shallower points have an increasingly transparent appearance
                # Deeper points have an increasingly solid appearance
                # Points with uniform depth have a transparent appearance
                sats = np.clip((zs - min(zs)) / _data_scale, 0.3, 1)

            else:
                raise NotImplementedError('Specified "shading_mode" must be 0 (default), 1, 2, or 3')

    rgba = np.broadcast_to(mcolors.to_rgba_array(colors), (len(zs), 4))

    # Change the alpha values of the colors using the generated alpha
    # multipliers
    return np.column_stack([rgba[:, :3], rgba[:, 3] * sats])
```

I would just have to add the kwarg to upstream function calls to give the user access.
Please let me know if there's a better/preferred way to handle when the shading mode is incorrectly specified (the NotImplementedError).